README.md
# repo2kaniko

[![Build Status](https://github.com/manics/repo2kaniko/actions/workflows/build.yml/badge.svg)](https://github.com/manics/repo2kaniko/actions/workflows/build.yml)
[![Container repository](https://img.shields.io/badge/quay.io-container-purple)](https://quay.io/repository/manics/repo2kaniko?tab=tags)

`repo2kaniko` is a plugin for [repo2docker](http://repo2docker.readthedocs.io) that lets you use [Kaniko](https://github.com/GoogleContainerTools/kaniko) instead of Docker.

Kaniko must run in a container, can only build images, and can only store images in a registry.
It has the big advantage that it i completely unprivileged and doesn't require any host configuration, unlike e.g. Podman which requires your system to have the correct cgroups configuration.

It does not use a local Docker store, so it is not possible to separate the build and push steps.
**This means the repo2docker `--push`/`--no-push` arguments have no effect.**
The default in this plugin is to automatically push the image to a registry after it is built.
Set Traitlet `KanikoEngine.push_image=False` or environment variable `KANIKO_PUSH_IMAGE=0` to disable pushing the image.

Kaniko does not cache layers locally, instead it uses a registry for caching.
You should probably use a dedicated local private registry for speed and not the remote registry used for storing the built image.
Use `KanikoEngine.cache_registry_credentials` or `KANIKO_CACHE_REGISTRY_CREDENTIALS` to specify the credentials for the cache registry.
These behave similarly to `ContainerEngine.registry_credentials` and `CONTAINER_ENGINE_REGISTRY_CREDENTIALS`.

## Running

```
podman run -it --rm quay.io/manics/repo2kaniko \
    repo2docker --debug --engine=kaniko \
    --Repo2Docker.user_id=1000 --user-name=jovyan \
    --KanikoEngine.registry_credentials=registry=quay.io \
    --KanikoEngine.registry_credentials=username=quay-user \
    --KanikoEngine.registry_credentials=password=quay-password \
    --image-name=quay.io/quay-user/r2d-test \
    --no-run \
    https://github.com/binderhub-ci-repos/minimal-dockerfile
```

With a local registry that can be used as a cache:

```
./run-local-registry.sh
REGISTRY=...
podman run -it --rm quay.io/manics/repo2kaniko \
    repo2docker --debug --engine=kaniko \
    --Repo2Docker.user_id=1000 --user-name=jovyan \
    --KanikoEngine.cache_registry=$REGISTRY/cache \
    --KanikoEngine.cache_registry_credentials=username=user \
    --KanikoEngine.cache_registry_credentials=password=password \
    --KanikoEngine.cache_registry_insecure=true \
    --image-name=$REGISTRY/r2d-test \
    --no-run \
    https://github.com/binderhub-ci-repos/minimal-dockerfile
```

## BinderHub

You can use repo2kaniko with a recent version of BinderHub
([1.0.0-0.dev.git.3350.he7995d6 or later](https://hub.jupyter.org/helm-chart/#development-releases-binderhub))
by adding the following to your configuration:

```yaml
config:
  KubernetesBuildExecutor:
    docker_host:
    build_image: "quay.io/manics/repo2kaniko:<tag>"
    repo2docker_extra_args:
      - --engine=kaniko
      - --debug

imageCleaner:
  enabled: false
```

If you want to use a local registry as a cache use something like this:

```yaml
config:
  KubernetesBuildExecutor:
    repo2docker_extra_args:
      - --engine=kaniko
      - --debug
      - --KanikoEngine.cache_registry=cache-registry-docker-registry:5000/cache
      - --KanikoEngine.cache_registry_credentials=username=user
      - --KanikoEngine.cache_registry_credentials=password=password
      - --KanikoEngine.cache_registry_insecure=true
```


CHANGELOG.md
# 0.1.0 - 2023-11-11


ci/build-and-test.sh
#!/usr/bin/env bash
set -eu

SCRIPTDIR=${BASH_SOURCE%/*}
cd "$SCRIPTDIR/.."

REPO2KANIKO_IMAGE=repo2kaniko:ci

REGISTRY=${REGISTRY_HOST:-}
if [ -z "$REGISTRY" ]; then
  # Use public IP when inside a container
  REGISTRY=$(hostname -I | awk '{print $1}'):5000
fi

PUSHED_IMAGE=$REGISTRY/test-conda:ci

if command -v podman; then
  ENGINE=podman
elif command -v docker; then
  ENGINE=docker
else
  echo "Neither podman nor docker are installed"
  exit 1
fi

echo "::group::Build repo2kaniko image"
$ENGINE build -t "$REPO2KANIKO_IMAGE" ./
echo "::endgroup::"

echo "::group::Run repo2kaniko"
$ENGINE run --rm --network=host -v "$PWD/ci/test-conda:/test-conda:ro,z" \
  "$REPO2KANIKO_IMAGE" repo2docker --engine kaniko --no-run --debug \
  --user-id=1000 --user-name=jovyan --image-name $PUSHED_IMAGE \
  --KanikoEngine.registry_credentials=registry=$REGISTRY \
  --KanikoEngine.registry_credentials=username=user \
  --KanikoEngine.registry_credentials=password=password \
  --KanikoEngine.registry_credentials=tls-verify=false \
  --KanikoEngine.cache_registry=$REGISTRY/cache \
  /test-conda
echo "::endgroup::"

echo "::group::Check repo2kaniko"
echo password | $ENGINE login --username=user --password-stdin --tls-verify=false localhost:5000
$ENGINE pull localhost:5000/test-conda:ci
echo "repo2docker --version:"
$ENGINE run $REPO2KANIKO_IMAGE repo2docker --version
echo "/home/jovyan/verify:"
$ENGINE run --rm localhost:5000/test-conda:ci /home/jovyan/verify

./ci/check-registry.py
echo "::endgroup::"


ci/check-registry.py
#!/usr/bin/env python

import os
import requests


REGISTRY_URL = "http://" + os.getenv("REGISTRY_HOST", "localhost:5000")
AUTH = ("user", "password")

r = requests.get(f"{REGISTRY_URL}/v2/_catalog", auth=AUTH)
r.raise_for_status()
repositories = r.json()["repositories"]
print(f"Repositories: {repositories}")
assert sorted(repositories) == ["cache", "test-conda"]

r = requests.get(f"{REGISTRY_URL}/v2/test-conda/tags/list", auth=AUTH)
r.raise_for_status()
test_conda_tags = r.json()["tags"]
print(f"test-conda tags: {test_conda_tags}")
assert test_conda_tags == ["ci"]

r = requests.get(f"{REGISTRY_URL}/v2/cache/tags/list", auth=AUTH)
r.raise_for_status()
cache_tags = r.json()["tags"]
print(f"cache tags: {cache_tags}")
# Should all be hashes
assert all(len(c) == 64 for c in cache_tags)
assert len(cache_tags) > 4


kaniko-runner/main.go
package main

import (
	"bufio"
	"encoding/base64"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"log"
	"net"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

type inputRequest struct {
	Command     []string           `json:"command"`
	Credentials []inputCredentials `json:"credentials"`
}

type inputCredentials struct {
	Registry string `json:"registry"`
	// Either set Auth (base64 string), or Username + Password
	Auth     string `json:"auth"`
	Username string `json:"username"`
	Password string `json:"password"`
	Insecure bool   `json:"insecure"`
}

//	{
//	  "auths": {
//	     "host": {
//	       "auth": "base64"
//	     }
//		 }
//	}
type dockerConfigAuths struct {
	Auths map[string]dockerConfigAuth `json:"auths"`
}

type dockerConfigAuth struct {
	Auth string `json:"auth"`
}

func listen(addr string) (net.Listener, error) {
	parts, err := url.Parse(addr)
	if err != nil {
		return nil, err
	}

	if parts.Scheme == "tcp" {
		log.Printf("Listening on tcp://%s", parts.Host)
		return net.Listen("tcp", parts.Host)
	} else if parts.Scheme == "unix" {
		log.Printf("Listening on unix://%s%s", parts.Host, parts.Path)
		return net.Listen("unix", parts.Host+parts.Path)
	} else {
		return nil, errors.New("unsupported network type")
	}
}

func returnError(conn net.Conn, err error) {
	fmt.Fprintln(conn, err)
	log.Println(err)
	fmt.Fprintln(conn, "status: FAILED")
}

func dockerConfigPath() string {
	executable, err := os.Executable()
	if err != nil {
		panic(err)
	}
	executableDir := filepath.Dir(executable)
	return filepath.Join(executableDir, ".docker", "config.json")
}

func loadDockerConfig() (dockerConfigAuths, error) {
	var config dockerConfigAuths
	configPath := dockerConfigPath()
	if _, err := os.Stat(configPath); err == nil {
		configFile, err := os.Open(configPath)
		if err != nil {
			return config, err
		}
		defer configFile.Close()

		if err := json.NewDecoder(configFile).Decode(&config); err != nil {
			return config, err
		}
	}
	if config.Auths == nil {
		config.Auths = make(map[string]dockerConfigAuth)
	}
	return config, nil
}

func saveDockerConfig(config dockerConfigAuths) error {
	configPath := dockerConfigPath()
	configDir := filepath.Dir(configPath)
	if err := os.MkdirAll(configDir, 0700); err != nil {
		return err
	}

	configFile, err := os.Create(configPath)
	if err != nil {
		return err
	}
	defer configFile.Close()

	encoder := json.NewEncoder(configFile)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(config); err != nil {
		return err
	}

	return nil
}

func handleConnection(conn net.Conn) error {
	defer conn.Close()

	// Read the incoming command from the client
	inputJson, err := bufio.NewReader(conn).ReadString('\n')
	if err != nil {
		return err
	}

	// input is a JSON object like:
	// {
	//   "command": ["string", "string", ...],
	//   "credentials": [{
	//     "registry": "https://index.docker.io/v1/",
	//		 "username": "username",
	//		 "password": "password",
	//		 "insecure": false
	//   },
	//   ...]
	// }
	var input inputRequest
	if err := json.Unmarshal([]byte(inputJson), &input); err != nil {
		returnError(conn, err)
		return err
	}

	dockerConfig, err := loadDockerConfig()

	if len(input.Credentials) > 0 {
		// Load the existing docker config
		if err != nil {
			returnError(conn, err)
			return err
		}

		for _, cred := range input.Credentials {
			// Add the credentials to the docker config
			if cred.Registry == "" {
				err := errors.New("registry is required")
				returnError(conn, err)
				return err
			}
			if cred.Auth != "" && (cred.Username != "" || cred.Password != "") {
				err := errors.New("cannot specify both auth and username/password")
				returnError(conn, err)
				return err
			}
			if cred.Auth != "" {
				dockerConfig.Auths[cred.Registry] = dockerConfigAuth{Auth: cred.Auth}
			} else {
				auth := base64.StdEncoding.EncodeToString([]byte(cred.Username + ":" + cred.Password))
				dockerConfig.Auths[cred.Registry] = dockerConfigAuth{Auth: auth}
			}
		}
	}

	// log the registries from dockerConfig
	var registries []string
	for registry := range dockerConfig.Auths {
		registries = append(registries, registry)
	}
	log.Printf("Registries: %s", strings.Join(registries, " "))

	if err := saveDockerConfig(dockerConfig); err != nil {
		returnError(conn, err)
		return err
	}

	log.Printf("Running command: %s", input.Command)

	// Run the command
	cmd := exec.Command(input.Command[0], input.Command[1:]...)

	// Get the output of the command
	stdout, erro := cmd.StdoutPipe()
	if erro != nil {
		returnError(conn, erro)
		return erro
	}
	stderr, erre := cmd.StderrPipe()
	if erre != nil {
		returnError(conn, erre)
		return erre
	}

	// Create a multiwriter that writes to both the console and the connection
	stdoutWriter := io.MultiWriter(os.Stdout, conn)
	stderrWriter := io.MultiWriter(os.Stderr, conn)

	if err := cmd.Start(); err != nil {
		returnError(conn, err)
		return err
	}

	// Use goroutines to handle stdout and stderr separately
	go io.Copy(stdoutWriter, stdout) //nolint:errcheck
	go io.Copy(stderrWriter, stderr) //nolint:errcheck

	if err = cmd.Wait(); err != nil {
		returnError(conn, err)
		return err
	}
	fmt.Fprintln(conn, "status: SUCCESS")
	return nil
}

func main() {
	var listenAddr string
	flag.StringVar(&listenAddr, "address", "tcp://localhost:8080", "address to listen on, e.g. unix:///tmp/go-runner.sock, tcp://localhost:8080")
	var multiple bool
	flag.BoolVar(&multiple, "multiple", false, "Keep listening after request finishes, , not supported by Kaniko")
	flag.Parse()

	ln, err := listen(listenAddr)
	if err != nil {
		panic(err)
	}

	for {
		conn, err := ln.Accept()
		if err != nil {
			panic(err)
		}

		if multiple {
			go handleConnection(conn) //nolint:errcheck
		} else {
			err := handleConnection(conn)
			// Give the client some time to read the output before exiting
			time.Sleep(2 * time.Second)
			if err != nil {
				log.Fatalf("ERROR: %v", err)
			}
			break
		}
	}
}


repo2kaniko/kaniko.py
# Use Kaniko instead of Docker
import json
import os
from shutil import copytree
import socket
import tarfile
from tempfile import TemporaryDirectory
from traitlets import default, Bool, Dict, Unicode, validate
from urllib.parse import urlparse

from repo2docker.engine import (
    ContainerEngine,
)

from repo2podman.podman import (
    log_debug,
    log_info,
    execute_cmd,
    exec_podman,
    exec_podman_stream,
)


class KanikoEngine(ContainerEngine):
    """
    Kaniko build engine
    """

    kaniko_executable = Unicode(
        "/kaniko/executor",
        help="The kaniko executable to use for all commands.",
        config=True,
    )

    kaniko_address = Unicode(
        # "tcp://localhost:8080",
        "",
        help=(
            "The address to connect to the Kaniko server. "
            "Set to empty to use the local Kaniko executor (not officially supported)."
        ),
        config=True,
    )

    kaniko_build_path = Unicode(
        "/workspace",
        help="Shared directory for build context when a separate Kaniko container is used.",
        config=True,
    )

    @validate("kaniko_build_path")
    def _validate_kaniko_build_path(self, proposal):
        value = proposal["value"]
        if not value.endswith("/"):
            value += "/"
        return value

    login_executable = Unicode(
        "skopeo",
        help="The executable to use for registry login.",
        config=True,
    )

    kaniko_loglevel = Unicode("", help="Kaniko log level", config=True)

    push_image = Bool(
        help="Push built image, default true.",
        config=True,
    )

    @default("push_image")
    def _push_image_default(self):
        """
        Set push_image from KANIKO_PUSH_IMAGE
        """
        return os.getenv("KANIKO_PUSH_IMAGE", "1").lower() in ("1", "t", "true")

    cache_registry = Unicode(
        "",
        help="Use this image registry as a cache for the build, e.g. 'localhost:5000/cache'.",
        config=True,
    )

    cache_registry_insecure = Bool(
        False,
        help="Allow insecure connections to the cache registry.",
        config=True,
    )

    cache_registry_credentials = Dict(
        help="""
        Credentials dictionary, if set will be used to authenticate with
        the cache registry. Typically this will include the keys:

            - `username`: The registry username
            - `password`: The registry password or token

        This can also be set by passing a JSON object in the
        KANIKO_CACHE_REGISTRY_CREDENTIALS environment variable.
        """,
        config=True,
    )

    @default("cache_registry_credentials")
    def _cache_registry_credentials_default(self):
        """
        Set the registry credentials from KANIKO_CACHE_REGISTRY_CREDENTIALS
        """
        obj = os.getenv("KANIKO_CACHE_REGISTRY_CREDENTIALS")
        if obj:
            try:
                return json.loads(obj)
            except json.JSONDecodeError:
                self.log.error("KANIKO_CACHE_REGISTRY_CREDENTIALS is not valid JSON")
                raise
        return {}

    cache_dir = Unicode(
        "",
        help=(
            "Read-only directory pre-populated with base images. "
            "It is not used for caching layers. See "
            "https://github.com/GoogleContainerTools/kaniko/tree/v1.17.0#caching-base-images"
        ),
        config=True,
    )

    def __init__(self, *, parent):
        super().__init__(parent=parent)

        lines = exec_podman(["version"], capture="stdout", exe=self.kaniko_executable)
        log_debug(lines)

    def _create_build_context(self, fileobj, srcpath, builddir):
        if fileobj:
            tarf = tarfile.open(fileobj=fileobj)
            tarf.extractall(builddir)
            log_debug(builddir)

            lines = execute_cmd(["ls", "-lRa", builddir], capture="stdout")
            log_debug(lines)
        elif srcpath:
            copytree(srcpath, builddir, symlinks=True)
        else:
            raise ValueError("No fileobj or srcpath")

    def _run_external_kaniko(self, cmdargs):
        input = {
            "command": [self.kaniko_executable] + cmdargs,
            "credentials": [],
        }
        # Dict with fields
        # - registry
        # - Either username and password
        # - Or auth
        if self.registry_credentials:
            input["credentials"].append(self.registry_credentials)
        if self.cache_registry_credentials:
            input["credentials"].append(self.cache_registry_credentials)

        url = urlparse(self.kaniko_address)
        if url.scheme == "tcp":
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            s.connect((url.hostname, url.port))
        elif url.scheme == "unix":
            s = socket.socket(socket.AF_UNIX)
            s.connect(url.path)

        message = json.dumps(input) + "\n"
        s.sendall(message.encode())

        # Receive data
        data = b""
        while True:
            chunk = s.recv(1024000)
            # print(chunk.decode(), end="")
            yield chunk.decode()
            if not chunk:
                break
            data += chunk

        s.close()

    def build(
        self,
        *,
        buildargs=None,
        cache_from=None,
        container_limits=None,
        tag="",
        custom_context=False,
        dockerfile="",
        fileobj=None,
        path="",
        labels=None,
        platform=None,
        **kwargs,
    ):
        log_debug("kaniko executor")

        cmdargs = []

        bargs = buildargs or {}
        for k, v in bargs.items():
            cmdargs.extend(["--build-arg", "{}={}".format(k, v)])

        if cache_from:
            log_info(f"Ignoring cache_from={cache_from}")

        if container_limits:
            log_info(f"Ignoring container_limits={container_limits}")

        if tag:
            cmdargs.extend(["--destination", tag])

        if dockerfile:
            cmdargs.extend(["--dockerfile", dockerfile])

        if labels:
            for k, v in labels.items():
                cmdargs.extend(["--label", "{}={}".format(k, v)])

        if platform:
            cmdargs.extend(["--custom-platform", platform])

        # TODO: what to do with these?
        # for ignore in ("custom_context", "decode"):
        #     try:
        #         kwargs.pop(ignore)
        #     except KeyError:
        #         pass

        if kwargs:
            raise ValueError("Additional kwargs not supported")

        ## Kaniko specific args

        # Kaniko uses a registry as a cache
        if self.cache_registry:
            cache_registry_host = self.cache_registry.split("/")[0]
            cmdargs.extend(
                [
                    "--cache=true",
                    "--cache-copy-layers=true",
                    "--cache-run-layers=true",
                    f"--cache-repo={self.cache_registry}",
                ]
            )
            if self.cache_registry_credentials:
                cache_credentials = self.cache_registry_credentials.copy()
                if self.cache_registry_insecure:
                    cmdargs.append(f"--insecure-registry={cache_registry_host}")
                    cache_credentials["tls-verify=false"] = None
                if not cache_credentials.get("registry"):
                    cache_credentials["registry"] = cache_registry_host
                self._login(**cache_credentials)

        if self.cache_dir:
            cmdargs.append(f"--cache-dir={self.cache_dir}")

        # Kaniko builds and pushes in one command
        if self.push_image:
            if self.registry_credentials:
                self._login(**self.registry_credentials)
        else:
            cmdargs.append("--no-push")

        # Avoid try-except so that if build errors occur they don't result in a
        # confusing message about an exception whilst handling an exception
        if self.kaniko_address:
            with TemporaryDirectory(prefix=self.kaniko_build_path) as builddir:
                self._create_build_context(fileobj, path, builddir)
                cmdargs.extend(["--context", builddir])
                for line in exec_podman_stream(cmdargs, exe=self.kaniko_executable):
                    yield line
        else:
            with TemporaryDirectory() as builddir:
                self._create_build_context(fileobj, path, builddir)
                cmdargs.extend(["--context", builddir])
                for line in exec_podman_stream(cmdargs, exe=self.kaniko_executable):
                    yield line

    def images(self):
        log_debug("kaniko images not supported")
        return []

    def inspect_image(self, image):
        raise NotImplementedError("kaniko inspect_image not supported")

    def _login(self, **kwargs):
        # kaniko doesn't support login, docker CLI doesn't support insecure, so use skopeo
        args = ["login"]

        registry = None
        password = None
        authfile = None

        for k, v in kwargs.items():
            if k == "password":
                password = v
            elif k == "registry":
                registry = v
            elif v is None:
                args.append(f"--{k}")
            else:
                args.append(f"--{k}={v}")

        if password is not None:
            args.append("--password-stdin")

        if authfile is None:
            authfile = os.path.expanduser("~/.docker/config.json")
        args.append(f"--authfile={authfile}")

        if registry is not None:
            args.append(registry)

        log_debug(f"{self.login_executable} login to registry {registry}")
        podman_kwargs = {"capture": "both"}
        if password is not None:
            podman_kwargs["input"] = password
        o = exec_podman(args, exe=self.login_executable, **podman_kwargs)
        log_debug(o)

    def push(self, image_spec):
        if not self.push_image:
            raise ValueError("Image must be pushed by setting push_image=True")
        # Otherwise should already have pushed as part of build
        return []

    def run(
        self,
        *args,
        **kwargs,
    ):
        raise NotImplementedError("kaniko run not supported")
